### DISCUSSION TITLE
Install issues

### DISCUSSION THREAD

**Discussion by gkhboiler** [dtcenter/METplus#1172] [Installation] [answered]
Hello,

I am attempting to install MET on a linux machine and seem to have successfully installed the libraries. When it gets to the MET code itself it seems to be failing on because of the python libraries but I am not sure why. I am getting a series of undefined references such as 
/p/home/hunterg/MET-10/met-10.0.0/src/libcode/vx_data2d_python/python_dataplane.cc:136: undefined reference to `PyErr_Occurred'
/p/home/hunterg/MET-10/met-10.0.0/src/libcode/vx_data2d_python/python_dataplane.cc:165: undefined reference to `PySys_SetArgv'
/p/home/hunterg/MET-10/met-10.0.0/src/libcode/vx_data2d_python/python_dataplane.cc:173: undefined reference to `PyImport_ImportModule'
/p/home/hunterg/MET-10/met-10.0.0/src/libcode/vx_data2d_python/python_dataplane.cc:181: undefined reference to `PyImport_ReloadModule'
/p/home/hunterg/MET-10/met-10.0.0/src/libcode/vx_data2d_python/python_dataplane.cc:185: undefined reference to `PyErr_Occurred'
I must have a setting wrong but I haven't been able to figure it out.

I am attaching my log files and my install scripts so you can see what am doing.

Thanks for your time,
Glenn
[make.log](https://github.com/dtcenter/METplus/files/7228629/make.log)
[configure.log](https://github.com/dtcenter/METplus/files/7228630/configure.log)
[config.log](https://github.com/dtcenter/METplus/files/7228631/config.log)
[compile_MET_all_sh.txt](https://github.com/dtcenter/METplus/files/7228638/compile_MET_all_sh.txt)
[install_MET_env_machine_intel.txt](https://github.com/dtcenter/METplus/files/7228646/install_MET_env_machine_intel.txt)

**Comment by JohnHalleyGotway** [comment, marked as answer]
Thanks for sending along your configuration and make log files. I see you're having trouble compiling python embedding for MET. I'm going to try to help you debug this, but may need to refer you to @jprestop if we get stuck.

First, when compiling MET, it's the ${MET_PYTHON_CC} and ${MET_PYTHON_LD} environment variables that are important. They are discussed in a [section of Appendix F](https://met.readthedocs.io/en/latest/Users_Guide/appendixF.html?highlight=MET_PYTHON_CC#compiling-python-support) of the MET User's Guide.

They are defined like this in the "install_MET_env_machine_intel.txt" that you sent:
```
export MET_PYTHON_LD="$(python3-config --ldflags)"
export MET_PYTHON_CC="$(python3-config --cflags)"
```
And checking in the "config.log" file you sent, their values are:
```
MET_PYTHON_CC='-I/p/app/CSE/CSEIntel.20210325/Release/anaconda3-2020.11/include/python3.8 -I/p/app/CSE/CSEIntel.20210325/Release/anaconda3-2020.11/include/python3.8  -Wno-unused-result -Wsign-compare -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O3 -ffunction-sections -pipe -isystem /p/app/CSE/CSEIntel.20210325/Release/anaconda3-2020.11/include -fdebug-prefix-map=/tmp/build/80754af9/python_1599203911753/work=/usr/local/src/conda/python-3.8.5 -fdebug-prefix-map=/app/CSE/CSEIntel.20210325/Release/anaconda3-2020.11=/usr/local/src/conda-prefix -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -flto -DNDEBUG -fwrapv -O3 -Wall'
MET_PYTHON_LD='-L/p/app/CSE/CSEIntel.20210325/Release/anaconda3-2020.11/lib/python3.8/config-3.8-x86_64-linux-gnu -L/p/app/CSE/CSEIntel.20210325/Release/anaconda3-2020.11/lib  -lcrypt -lpthread -ldl  -lutil -lrt -lm -lm '
```
You'll notice here that there's lots and lots of settings in MET_PYTHON_CC, and it's possible that all those extra compilation options are wreaking some havoc. I'd recommend trying the following... Comment out the existing MET_PYTHON_CC setting and define it manually. So edit "install_MET_env_machine_intel.txt" like this:
```
#Python cflags created using python3-config
#export MET_PYTHON_CC="$(python3-config --cflags)"
export MET_PYTHON_CC="-I/p/app/CSE/CSEIntel.20210325/Release/anaconda3-2020.11/include/python3.8"
echo $MET_PYTHON_CC
```
Then reconfigure/recompile MET using these settings.

Does that produce any better result?

**Reply by jprestop** [reply]
Thanks, @JohnHalleyGotway!  I was just looking over these files as well.  I was going to make the same suggestion for MET_PYTHON_CC.  I also noticed one more thing.  MET_PYTHON_LD is missing "-lpython3.8", which is needed.  Please also edit "install_MET_env_machine_intel.txt" like this:

```
#Python ldflags created using python3-config                                                                                                                    
#export MET_PYTHON_LD="$(python3-config --ldflags)"
export MET_PYTHON_LD="-L/p/app/CSE/CSEIntel.20210325/Release/anaconda3-2020.11/lib/python3.8/config-3.8-x86_64-linux-gnu -L/p/app/CSE/CSEIntel.20210325/Release/anaconda3-2020.11/lib -lpython3.8 -lcrypt -lpthread -ldl  -lutil -lrt -lm"
echo $MET_PYTHON_LD
```

Unfortunately, running "python3-config --cflags" and "python3-config --ldflags" won't always give us the desired results and can have extra settings or not enough settings.

Please give that a try and let us know how it goes.

**Reply by gkhboiler** [reply]
Thanks to both of you for the recommendations. I used them both and they worked! My only other mini trip-up was that I had FC set to gfortran in my login script so the compile script kept FC as gfortran while the rest of the compilers settings were being set to intel which caused an incompatability. When I reset FC to "" then all of the compilers were set correctly and I received a
Finished compiling at Tue Sep 28 12:07:39 EDT 2021

Thank you!!!

**Reply by JohnHalleyGotway** [reply]
Excellent! Thanks for following up to confirm.

At this point I'm going to LOCK this conversation to prevent future posts. Our team has decided that once discussions have been answered, and the answer has been confirmed, we'll lock them. We want to encourage users to ask new questions in new discussions rather than posting to old ones. Hopefully that'll make the questions/answers easier for other users to follow.

So if/when more issues/questions arise, please feel free to start a new discussion.
